File: mac-app/TipUITests/tipUITests.m
//
//  test2UITests.m
//  test2UITests
//
//  Created by Tanin Na Nakorn on 12/28/19.
//  Copyright © 2019 Tanin Na Nakorn. All rights reserved.
//

#import <XCTest/XCTest.h>

@interface tipUITests : XCTestCase

@end

@implementation tipUITests

NSPasteboard* pboard;
XCUIApplication* app;

- (void)setUp {
    // Put setup code here. This method is called before the invocation of each test method in the class.

    // In UI tests it is usually best to stop immediately when a failure occurs.
    self.continueAfterFailure = NO;

    // In UI tests it’s important to set the initial state - such as interface orientation - required for your tests before they run. The setUp method is a good place to do this.
    pboard = [NSPasteboard generalPasteboard];
    [pboard clearContents];
    [pboard setString:@"" forType:NSPasteboardTypeString];
    
    app = [[XCUIApplication alloc] init];
    
}

- (void) launchWithProvider:(NSString *)provider {
    NSBundle *bundle = [NSBundle bundleForClass:[self class]];
    NSString *path = [bundle pathForResource:provider ofType:@"rb"];
    app.launchArguments = @[@"-test", @"TestInput", @"-provider", path];
    [app launch];
}

- (void)tearDown {
    [app terminate];
}

- (void)testClickingOnText {
    [self launchWithProvider:@"provider"];
    [app.popovers.element.tableRows.firstMatch.cells.firstMatch click];
    XCTAssert([@"Return TestInput" isEqualToString:[pboard stringForType:NSPasteboardTypeString]]);
}

- (void)testClickingOnUrl {
    [self launchWithProvider:@"provider"];
    [[app.popovers.element.tableRows elementBoundByIndex:1].cells.firstMatch click];
    [NSThread sleepForTimeInterval:0.1f];
    XCTAssert([@"tanintip://TestInput" isEqualToString:[pboard stringForType:NSPasteboardTypeString]]);
}

- (void)testNoTips {
    [self launchWithProvider:@"empty_provider"];
    XCTAssert([@"No tips. You can add tips through your provider script. Click to see the instruction." isEqualToString:[[app.popovers childrenMatchingType:XCUIElementTypeAny] elementBoundByIndex:1].firstMatch.value]);
    
    [app.popovers.element click];
    [NSThread sleepForTimeInterval:0.1f];
    XCTAssert([@"OpenProviderInstruction" isEqualToString:[pboard stringForType:NSPasteboardTypeString]]);
}

- (void)testNoProvider {
    app.launchArguments = @[@"-test", @"TestInput", @"-provider", @"/tmp/some-non-existent-provider.rb"];
    [app launch];
    XCTAssert([@"/tmp/some-non-existent-provider.rb doesn't exist. Please make a provider script. Click to see instruction." isEqualToString:[[app.popovers childrenMatchingType:XCUIElementTypeAny] elementBoundByIndex:1].firstMatch.value]);
    
    [app.popovers.element click];
    [NSThread sleepForTimeInterval:0.1f];
    XCTAssert([@"OpenProviderInstruction" isEqualToString:[pboard stringForType:NSPasteboardTypeString]]);
}

- (void)testUnexecutableProvider {
    [self launchWithProvider:@"unexecutable_provider"];
    XCTAssert([(NSString*)[[app.popovers childrenMatchingType:XCUIElementTypeAny] elementBoundByIndex:1].firstMatch.value containsString:@"Provider isn't executable. Please chmod 755"]);
    
    [app.popovers.element click];
    [NSThread sleepForTimeInterval:0.1f];
    XCTAssert([@"None" isEqualToString:[pboard stringForType:NSPasteboardTypeString]]);
}

- (void)testMalformedJson {
    [self launchWithProvider:@"malformed_json_provider"];
    XCTAssert([@"Malformed JSON returned from provider. Click to see logs in Console. You'll need to set the filter Process=Tip." isEqualToString:[[app.popovers childrenMatchingType:XCUIElementTypeAny] elementBoundByIndex:1].firstMatch.value]);
    
    [app.popovers.element click];
    [NSThread sleepForTimeInterval:0.1f];
    XCTAssert([@"OpenConsole" isEqualToString:[pboard stringForType:NSPasteboardTypeString]]);
}

- (void)testError {
    [self launchWithProvider:@"error_provider"];
    XCTAssert([@"Error occurred. Click to see logs in Console. You'll need to set the filter Process=Tip." isEqualToString:[[app.popovers childrenMatchingType:XCUIElementTypeAny] elementBoundByIndex:1].firstMatch.value]);
    
    [app.popovers.element click];
    [NSThread sleepForTimeInterval:0.1f];
    XCTAssert([@"OpenConsole" isEqualToString:[pboard stringForType:NSPasteboardTypeString]]);
}

- (void)testLaunchPerformance {
    if (@available(macOS 10.15, iOS 13.0, tvOS 13.0, *)) {
        // This measures how long it takes to launch your application.
        [self measureWithMetrics:@[XCTOSSignpostMetric.applicationLaunchMetric] block:^{
            [[[XCUIApplication alloc] init] launch];
        }];
    }
}

@end

File: mac-app/TipUITests/TipUITests.swift
import XCTest

final class TipUITests : XCTestCase {
    
    let pasteBoard = NSPasteboard.general
    let app = XCUIApplication()
    
    override func setUp() {
        pasteBoard.clearContents()
        pasteBoard.setString("", forType: .string)
    }
    
    override func tearDown() {
        app.terminate()
    }
    
    func testBundleId() {
        launch(withName: "bundle_id_provider")
        
        let components = getLabel(rowIndex: 0).split(separator: " ")
        XCTAssertEqual("Return", components[0])
        XCTAssertEqual("TestInput", components[1])
        XCTAssertEqual("--bundle-identifier", components[2])
        XCTAssertNotNil(components[3].range(of: "^[a-zA-Z0-9]+(\\.[a-zA-Z0-9]+)+$", options: .regularExpression))
    }
    
    func testGoodProviderClickingOnText() {
        launch(withName: "good_provider")
        XCTAssertEqual("Return TestInput", getLabel(rowIndex: 0))
        click(rowIndex: 0)
        XCTAssertEqual("Return TestInput", pasteBoard.string(forType: .string))
    }
    
    func testFirstRowIsSelectedAsDefault() {
        launch(withName: "good_provider")
        XCTAssertEqual("Return TestInput", getLabel(rowIndex: 0))
        app.typeKey(XCUIKeyboardKey.`return`, modifierFlags: XCUIElement.KeyModifierFlags.init())
        XCTAssertEqual("Return TestInput", pasteBoard.string(forType: .string))
    }
    
    func testGoodProviderClickingOnTextWithLabel() {
        launch(withName: "good_provider")
        XCTAssertEqual("{\n  \"input\": \"TestInput\",\n  \"other\": true\n}", getLabel(rowIndex: 1))
        click(rowIndex: 1)
        XCTAssertEqual("Value TestInput", pasteBoard.string(forType: .string))
    }
    
    func testGoodProviderClickingOnURL() {
        launch(withName: "good_provider")
        XCTAssertEqual("Go to TestInput", getLabel(rowIndex: 2))
        click(rowIndex: 2)
        XCTAssertEqual("tanintip://TestInput", pasteBoard.string(forType: .string))
    }
    
    func testGoodProviderClickingOnExecute() {
        launch(withName: "good_provider")
        
        XCTAssertEqual("Execute TestInput", getLabel(rowIndex: 3))
        click(rowIndex: 3)
        
        Thread.sleep(forTimeInterval: 1)
        XCTAssertEqual("Reexecute --execute test test2", getLabel(rowIndex: 1))
        click(rowIndex: 1)
        
        Thread.sleep(forTimeInterval: 1)
        
        XCTAssertEqual("Result --execute reexecute", getLabel(rowIndex: 0))
        click(rowIndex: 0)
        XCTAssertEqual("Result --execute reexecute", pasteBoard.string(forType: .string))
    }
    
    func testProviderAutoExecuteOnText() {
        launch(withName: "auto_execute_text_provider")
        usleep(useconds_t(1000 * 1000))
        XCTAssertEqual("Return Auto TestInput", pasteBoard.string(forType: .string))
    }
    
    func testProviderAutoExecuteOnUrl() {
        launch(withName: "auto_execute_url_provider", expectPopover: false)
        usleep(useconds_t(200 * 1000))
        XCTAssertEqual("tanintip://auto-TestInput", pasteBoard.string(forType: .string))
    }
    
    func testNoProvider() {
        app.launchArguments = ["-test", "TestInput", "-provider", "/tmp/something-doesn-exist.rb"]
        app.launch()
        XCTAssertEqual("/tmp/something-doesn-exist.rb doesn't exist. Please make a provider script. Click to see instruction.", app.popovers.children(matching: .any).element(boundBy: 1).firstMatch.value as! String)
        
        app.popovers.element.click()
        usleep(useconds_t(200 * 1000))
        XCTAssertEqual("OpenProviderInstruction", pasteBoard.string(forType: .string))
    }
    
    func testNoTip() {
        launch(withName: "empty_provider")
        XCTAssertEqual("No tips. You can add tips through your provider script. Click to see the instruction.", app.popovers.children(matching: .any).element(boundBy: 1).firstMatch.value as! String)
        
        app.popovers.element.click()
        usleep(useconds_t(200 * 1000))
        XCTAssertEqual("OpenProviderInstruction", pasteBoard.string(forType: .string))
    }
    
    func testProviderUnexecutable() {
        launch(withName: "unexecutable_provider")
        let value = app.popovers.children(matching: .any).element(boundBy: 1).firstMatch.value as! String
        let firstSlash = value.firstIndex(of: "/") ?? value.endIndex
        XCTAssertEqual("Provider isn't executable. Please chmod 755 ", value[..<firstSlash])
        
        app.popovers.element.click()
        usleep(useconds_t(200 * 1000))
        XCTAssertEqual("None", pasteBoard.string(forType: .string))
    }
    
    func testProvideMalformedJson() {
        launch(withName: "malformed_json_provider")
        XCTAssertEqual("Malformed JSON returned from provider. Click to see logs in Console. You'll need to set the filter Process=Tip.", app.popovers.children(matching: .any).element(boundBy: 1).firstMatch.value as! String)
        
        app.popovers.element.click()
        usleep(useconds_t(200 * 1000))
        XCTAssertEqual("OpenConsole", pasteBoard.string(forType: .string))
    }
    
    func testErrorProvider() {
        launch(withName: "error_provider")
        XCTAssertEqual("Error occurred. Click to see logs in Console. You'll need to set the filter Process=Tip.", app.popovers.children(matching: .any).element(boundBy: 1).firstMatch.value as! String)
        
        app.popovers.element.click()
        usleep(useconds_t(200 * 1000))
        XCTAssertEqual("OpenConsole", pasteBoard.string(forType: .string))
    }
    
    private func launch(withName: String, expectPopover: Bool = true){
        let file = Bundle(for: type(of: self)).path(forResource: withName, ofType: "rb")
        app.launchArguments = ["-test", "TestInput", "-provider", file!]
        app.launch()
        app.activate()
        if (expectPopover) {
            XCTAssertEqual(
                XCTWaiter.wait(for: [XCTKVOExpectation(keyPath:"exists", object: app.popovers.element, expectedValue: true)], timeout: 3),
                .completed)
        }
    }
    
    private func click(rowIndex: Int)  {
        let button = app.popovers.element.tableRows.element(boundBy: rowIndex).cells.element(boundBy: 0).firstMatch
        XCTAssertEqual(
            XCTWaiter.wait(for: [XCTKVOExpectation(keyPath:"isHittable", object: button, expectedValue: true)], timeout: 3),
            .completed)
        button.click()
    }
    
    private func getLabel(rowIndex: Int) -> String {
        let text = app.popovers.element.tableRows.element(boundBy: rowIndex).cells.element(boundBy: 1).staticTexts.element(boundBy: 0)
        XCTAssertEqual(
            XCTWaiter.wait(for: [XCTKVOExpectation(keyPath:"exists", object: text, expectedValue: true)], timeout: 3),
            .completed)
        return text.value as! String
    }
    
}
